### ISSUE TITLE
cannot compile for windows.

### ISSUE THREAD

**Issue by VinceLewis** [pokerth/pokerth#407]
Hi,

Trying to follow the instructions in build_mingw_windows.txt to compile for windows. 
After an initial try and failure, I've done some digging on the web, and this is my current build script:

export POKERTH_DIR=/home/vince/Documents/pokerth-1.1.2-rc
echo 'LIBS += -lmodplug' >> $POKERTH_DIR/pokerth_game.pro
echo 'LIBS += -lcrypt32' >> $POKERTH_DIR/pokerth_server.pro

export PATH="/opt/mxe/usr/bin:$PATH"
export TGT='i686-w64-mingw32.static'
rm -rf $POKERTH_DIR.$TGT.build
mkdir $POKERTH_DIR.$TGT.build
cd $POKERTH_DIR.$TGT.build

/opt/mxe/usr/bin/$TGT-qmake-qt4 $POKERTH_DIR/pokerth.pro DEFINES+=BOOST_THREAD_USE_LIB
make release

Initially it said could not find protoc, so I downloaded the latest release binary from github (3.12.3) and put it in /usr/local/bin

Now when I run the script it fails with this in the log:

/opt/mxe/usr/bin/i686-w64-mingw32.static-qmake-qt4 /home/vince/Documents/pokerth-1.1.2-rc/pokerth_protocol.pro DEFINES+=BOOST_THREAD_USE_LIB -o Makefile.pokerth_protocol
sh: 1: protoc: Permission denied
sh: 1: protoc: Permission denied
sh: 1: protoc: Permission denied
sh: 1: protoc: Permission denied
sh: 1: protoc: Permission denied
sh: 1: protoc: Permission denied
sh: 1: protoc: Permission denied
sh: 1: protoc: Permission denied
sh: 1: protoc: Permission denied
make -f Makefile.pokerth_protocol release
make[1]: Entering directory '/home/vince/Documents/pokerth-1.1.2-rc.i686-w64-mingw32.static.build'
make -f Makefile.pokerth_protocol.Release
make[2]: Entering directory '/home/vince/Documents/pokerth-1.1.2-rc.i686-w64-mingw32.static.build'
i686-w64-mingw32.static-g++ -c -pipe -fno-keep-inline-dllexport -O2 -Wall -Wextra -frtti -fexceptions -mthreads -DUNICODE -DBOOST_THREAD_USE_LIB -DENABLE_IPV6 -DCURL_STATICLIB -D_WIN32_WINNT=0x0501 -DQT_THREAD_SUPPORT -DQT_NO_DEBUG -I'/opt/mxe/usr/i686-w64-mingw32.static/qt/include' -I'../pokerth-1.1.2-rc' -I'../pokerth-1.1.2-rc/src' -I'/opt/mxe/usr/i686-w64-mingw32.static/qt/include/ActiveQt' -I'mocs' -I'.' -I'/opt/mxe/usr/i686-w64-mingw32.static/qt/mkspecs/win32-g++-4.6' -o obj/pokerth.pb.o ../pokerth-1.1.2-rc/src/third_party/protobuf/pokerth.pb.cc
In file included from ../pokerth-1.1.2-rc/src/third_party/protobuf/pokerth.pb.cc:5:0:
../pokerth-1.1.2-rc/src/third_party/protobuf/pokerth.pb.h:17:2: error: #error This file was generated by an older version of protoc which is
 #error This file was generated by an older version of protoc which is
  ^
../pokerth-1.1.2-rc/src/third_party/protobuf/pokerth.pb.h:18:2: error: #error incompatible with your Protocol Buffer headers. Please
 #error incompatible with your Protocol Buffer headers.  Please
  ^
../pokerth-1.1.2-rc/src/third_party/protobuf/pokerth.pb.h:19:2: error: #error regenerate this file with a newer version of protoc.
 #error regenerate this file with a newer version of protoc.
  ^
In file included from /opt/mxe/usr/lib/gcc/i686-w64-mingw32.static/5.5.0/include/c++/atomic:38:0,
                 from /opt/mxe/usr/i686-w64-mingw32.static/include/google/protobuf/generated_message_util.h:42,
                 from ../pokerth-1.1.2-rc/src/third_party/protobuf/pokerth.pb.h:22,
                 from ../pokerth-1.1.2-rc/src/third_party/protobuf/pokerth.pb.cc:5:
/opt/mxe/usr/lib/gcc/i686-w64-mingw32.static/5.5.0/include/c++/bits/c++0x_warning.h:32:2: error: #error This file requires compiler and library support for the ISO C++ 2011 standard. This support must be enabled with the -std=c++11 or -std=gnu++11 compiler options.
 #error This file requires compiler and library support \
  ^
../pokerth-1.1.2-rc/src/third_party/protobuf/pokerth.pb.cc:12:50: fatal error: google/protobuf/wire_format_lite_inl.h: No such file or directory
compilation terminated.
make[2]: *** [Makefile.pokerth_protocol.Release:123: obj/pokerth.pb.o] Error 1
make[2]: Leaving directory '/home/vince/Documents/pokerth-1.1.2-rc.i686-w64-mingw32.static.build'
make[1]: *** [Makefile.pokerth_protocol:34: release] Error 2
make[1]: Leaving directory '/home/vince/Documents/pokerth-1.1.2-rc.i686-w64-mingw32.static.build'
make: *** [Makefile:287: sub-pokerth_protocol-pro-sub_Release_ordered] Error 2

Any ideas?
I'm running Ubuntu 19.10

Addendum: When I try and compile for linux I get:
/usr/share/qt4/bin/qmake /home/vince/Documents/pokerth-1.1.2-rc/pokerth_protocol.pro -o Makefile.pokerth_protocol
sh: 1: protoc: Permission denied
sh: 1: protoc: Permission denied
sh: 1: protoc: Permission denied
make -f Makefile.pokerth_protocol 
make[1]: Entering directory '/home/vince/Documents/pokerth-1.1.2-rc'
g++ -c -m64 -pipe -O2 -Wall -W -D_REENTRANT -fPIC -DENABLE_IPV6 -DQT_NO_DEBUG -I/usr/share/qt4/mkspecs/linux-g++-64 -I. -I/usr/include/qt4 -I. -Isrc -I/include -Imocs -o obj/pokerth.pb.o src/third_party/protobuf/pokerth.pb.cc
In file included from src/third_party/protobuf/pokerth.pb.cc:5:
src/third_party/protobuf/pokerth.pb.h:9:10: fatal error: google/protobuf/stubs/common.h: No such file or directory
    9 | #include <google/protobuf/stubs/common.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[1]: *** [Makefile.pokerth_protocol:211: obj/pokerth.pb.o] Error 1
make[1]: Leaving directory '/home/vince/Documents/pokerth-1.1.2-rc'
make: *** [Makefile:41: sub-pokerth_protocol-pro-make_default-ordered] Error 2

**Comment by q4z1** [comment]
Hi,

I'm not sure about the "permission denied" errors - but for this `regenerate this file with a newer version of protoc` you will have to delete and recretate the files in `src/third_party/protobuf/` :

`protoc -I=/path/to/pokerth/ --cpp_out=/path/to/pokerth/src/third_party/protobuf/ /path/to/pokerth/pokerth.proto /path/to/pokerth/chatcleaner.proto`

Can you try that out?

Regards.

**Comment by VinceLewis** [comment]
The "permission denied" goes away if I use sudo, so I'm doing that.
I renamed the old files in the directory.
the command you suggested created new chatcleaner.pb.cc and
chatcleaner.pb.h files, but not the pokerth.pb* files.

After running my build script I'm now getting the error:
[libprotobuf WARNING
../../../../../src/google/protobuf/compiler/parser.cc:648] No syntax
specified for the proto file: pokerth.proto. Please use 'syntax =
"proto2";' or 'syntax = "proto3";' to specify a syntax version. (Defaulted
to proto2 syntax.)
pokerth.proto:704:26: Expected enum value or number range.
/opt/mxe/usr/bin/i686-w64-mingw32.static-qmake-qt4
/home/vince/Documents/pokerth-1.1.2-rc/pokerth_protocol.pro
DEFINES+=BOOST_THREAD_USE_LIB -o Makefile.pokerth_protocol
[libprotobuf WARNING
../../../../../src/google/protobuf/compiler/parser.cc:648] No syntax
specified for the proto file: pokerth.proto. Please use 'syntax =
"proto2";' or 'syntax = "proto3";' to specify a syntax version. (Defaulted
to proto2 syntax.)
pokerth.proto:704:26: Expected enum value or number range.
[libprotobuf WARNING
../../../../../src/google/protobuf/compiler/parser.cc:648] No syntax
specified for the proto file: chatcleaner.proto. Please use 'syntax =
"proto2";' or 'syntax = "proto3";' to specify a syntax version. (Defaulted
to proto2 syntax.)
[libprotobuf WARNING
../../../../../src/google/protobuf/compiler/parser.cc:648] No syntax
specified for the proto file: pokerth.proto. Please use 'syntax =
"proto2";' or 'syntax = "proto3";' to specify a syntax version. (Defaulted
to proto2 syntax.)
pokerth.proto:704:26: Expected enum value or number range.
[libprotobuf WARNING
../../../../../src/google/protobuf/compiler/parser.cc:648] No syntax
specified for the proto file: pokerth.proto. Please use 'syntax =
"proto2";' or 'syntax = "proto3";' to specify a syntax version. (Defaulted
to proto2 syntax.)
pokerth.proto:704:26: Expected enum value or number range.
[libprotobuf WARNING
../../../../../src/google/protobuf/compiler/parser.cc:648] No syntax
specified for the proto file: chatcleaner.proto. Please use 'syntax =
"proto2";' or 'syntax = "proto3";' to specify a syntax version. (Defaulted
to proto2 syntax.)
[libprotobuf WARNING
../../../../../src/google/protobuf/compiler/parser.cc:648] No syntax
specified for the proto file: pokerth.proto. Please use 'syntax =
"proto2";' or 'syntax = "proto3";' to specify a syntax version. (Defaulted
to proto2 syntax.)
pokerth.proto:704:26: Expected enum value or number range.
WARNING: Failure to find: src/third_party/protobuf/pokerth.pb.cc
WARNING: Failure to find: src/third_party/protobuf/pokerth.pb.h
[libprotobuf WARNING
../../../../../src/google/protobuf/compiler/parser.cc:648] No syntax
specified for the proto file: pokerth.proto. Please use 'syntax =
"proto2";' or 'syntax = "proto3";' to specify a syntax version. (Defaulted
to proto2 syntax.)
pokerth.proto:704:26: Expected enum value or number range.
[libprotobuf WARNING
../../../../../src/google/protobuf/compiler/parser.cc:648] No syntax
specified for the proto file: chatcleaner.proto. Please use 'syntax =
"proto2";' or 'syntax = "proto3";' to specify a syntax version. (Defaulted
to proto2 syntax.)
[libprotobuf WARNING
../../../../../src/google/protobuf/compiler/parser.cc:648] No syntax
specified for the proto file: pokerth.proto. Please use 'syntax =
"proto2";' or 'syntax = "proto3";' to specify a syntax version. (Defaulted
to proto2 syntax.)
pokerth.proto:704:26: Expected enum value or number range.
WARNING: Failure to find: src/third_party/protobuf/pokerth.pb.cc
WARNING: Failure to find: src/third_party/protobuf/pokerth.pb.h
make -f Makefile.pokerth_protocol release
make[1]: Entering directory
'/home/vince/Documents/pokerth-1.1.2-rc.i686-w64-mingw32.static.build'
make -f Makefile.pokerth_protocol.Release
make[2]: Entering directory
'/home/vince/Documents/pokerth-1.1.2-rc.i686-w64-mingw32.static.build'
make[2]: *** No rule to make target 'src/third_party/protobuf/pokerth.pb.cc',
needed by 'obj/pokerth.pb.o'. Stop.
make[2]: Leaving directory
'/home/vince/Documents/pokerth-1.1.2-rc.i686-w64-mingw32.static.build'
make[1]: *** [Makefile.pokerth_protocol:34: release] Error 2
make[1]: Leaving directory
'/home/vince/Documents/pokerth-1.1.2-rc.i686-w64-mingw32.static.build'
make: *** [Makefile:287: sub-pokerth_protocol-pro-sub_Release_ordered]
Error 2

**Comment by q4z1** [comment]
Then something went wrong with the re-creation - try it without chatcleaner.pro:

`protoc -I=/path/to/pokerth/ --cpp_out=/path/to/pokerth/src/third_party/protobuf/ /path/to/pokerth/pokerth.proto`

seeh [https://developers.google.com/protocol-buffers/docs/reference/cpp-generated](https://developers.google.com/protocol-buffers/docs/reference/cpp-generated)

The warnings `Please use 'syntax = "proto2";' or 'syntax = "proto3";` can be ignored.

**Comment by VinceLewis** [comment]
Still fails, same error. See below:
vince@vince-desktop:~/Documents/pokerth-1.1.2-rc$ sudo protoc
-I=/home/vince/Documents/pokerth-1.1.2-rc/
--cpp_out=/home/vince/Documents/pokerth-1.1.2-rc/src/third_party/protobuf/
/home/vince/Documents/pokerth-1.1.2-rc/pokerth.proto
[libprotobuf WARNING
../../../../../src/google/protobuf/compiler/parser.cc:648] No syntax
specified for the proto file: pokerth.proto. Please use 'syntax =
"proto2";' or 'syntax = "proto3";' to specify a syntax version. (Defaulted
to proto2 syntax.)
pokerth.proto:704:26: Expected enum value or number range.

Looks like a bug in pokerth.proto

**Comment by q4z1** [comment]
This related to this issue: #339

**Comment by VinceLewis** [comment]
so a standard "make" on linux, to try and build on linux, gives this:
make -f Makefile.pokerth_protocol
make[1]: Entering directory '/home/vince/Documents/pokerth-1.1.2-rc'
g++ -c -m64 -pipe -O2 -Wall -W -D_REENTRANT -fPIC -DENABLE_IPV6
-DQT_NO_DEBUG -I/usr/share/qt4/mkspecs/linux-g++-64 -I. -I/usr/include/qt4
-I. -Isrc -I/include -Imocs -o obj/pokerth.pb.o src/third_party/protobuf/
pokerth.pb.cc
In file included from src/third_party/protobuf/pokerth.pb.cc:4:
src/third_party/protobuf/pokerth.pb.h:10:10: fatal error:
google/protobuf/port_def.inc: No such file or directory
   10 | #include <google/protobuf/port_def.inc>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[1]: *** [Makefile.pokerth_protocol:211: obj/pokerth.pb.o] Error 1
make[1]: Leaving directory '/home/vince/Documents/pokerth-1.1.2-rc'
make: *** [Makefile:41: sub-pokerth_protocol-pro-make_default-ordered]
Error 2

and if I try the windows build on my linux box, by doing this:
/opt/mxe/usr/bin/$TGT-qmake-qt4 $POKERTH_DIR/pokerth.pro
DEFINES+=BOOST_THREAD_USE_LIB

I get the attached log file

Note: I started trying to compile the windows version because that's my
target platform. I've only been trying to compile on Linux first as a
stepping stone.

**Comment by q4z1** [comment]
If protobuf-compiler is already installed, try sudo apt-get remove protobuf-compiler than reinstall via apt-get install.

**Comment by q4z1** [comment]
It might be way easier to use `apt-get build-deps pokerth` as this will install quite all necessary dependencies for building pokerth - including the necessary protbuf libs.

Sources should be enabled in apt sources.list

You will also avoid using sudo for protoc, which is usually not a good idea.

**Comment by VinceLewis** [comment]
I did
sudo apt-get remove protobuf-compiler

then
sudo apt-get install protobuf-compiler

then enabled sources

Then tried
sudo apt-get build-dep pokerth
which worked fine, no errors

then tried linux make. Fails with same error:
make
make -f Makefile.pokerth_protocol
make[1]: Entering directory '/home/vince/Documents/pokerth-1.1.2-rc'
g++ -c -m64 -pipe -O2 -Wall -W -D_REENTRANT -fPIC -DENABLE_IPV6
-DQT_NO_DEBUG -I/usr/share/qt4/mkspecs/linux-g++-64 -I. -I/usr/include/qt4
-I. -Isrc -I/include -Imocs -o obj/pokerth.pb.o src/third_party/protobuf/
pokerth.pb.cc
In file included from src/third_party/protobuf/pokerth.pb.cc:4:
src/third_party/protobuf/pokerth.pb.h:10:10: fatal error:
google/protobuf/port_def.inc: No such file or directory
   10 | #include <google/protobuf/port_def.inc>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[1]: *** [Makefile.pokerth_protocol:211: obj/pokerth.pb.o] Error 1
make[1]: Leaving directory '/home/vince/Documents/pokerth-1.1.2-rc'
make: *** [Makefile:41: sub-pokerth_protocol-pro-make_default-ordered]
Error

Do you know where I should be able to find google/protobuf/port_def.inc?

I checked /usr/local/include/local and no joy, it looks like this:
[image: image.png]

Can you confirm which version of protobuf you are using please?
For reference:

sudo apt-cache policy protobuf-compiler
protobuf-compiler:
  Installed: 3.6.1.3-2
  Candidate: 3.6.1.3-2
  Version table:
 *** 3.6.1.3-2 500
        500 http://gb.archive.ubuntu.com/ubuntu eoan/universe amd64 Packages
        100 /var/lib/dpkg/status

**Comment by q4z1** [comment]
Did you remove all fromer installations? Symlinks might not trigger if you do not delete all former tries. I'd suggest a fresh clean Ubuntu.

**Comment by albmed** [comment]
A couple of comments: 

- Have you read [wiki](https://github.com/pokerth/pokerth/wiki/Building-PokerTH-for-Windows)? It contains updated information to build windows binaries
- Keep in mind you are using a cross-compiler (MXE) to build binaries. So, **you must use** `protoc` within you MXE installation and not the one in your linux system. In my system it's located in `/opt/mingw/usr/x86_64-unknown-linux-gnu/bin/protoc`. This also applies for `qmake` and `make`. 
- `sudo` is mandatory unless you are building using `root` user (which is IMHO a bad idea). `/opt` is a system folder and normal users (even admins) do not have permission to write in that folder, by default. 

Regards.

**Comment by VinceLewis** [comment]
@q4z1 - Alas I dont have a 2nd machine to do a fresh ubuntu install as previously suggested.
@albmed - so. Read the wiki. Copied the protobuf includes (eg google) into /usr/local/include otherwise my system didn't find it, even if I altered $path.

Tried just a **Linux** make. 

Got past this error:
In file included from src/third_party/protobuf/pokerth.pb.cc:4:
src/third_party/protobuf/pokerth.pb.h:10:10: fatal error:
google/protobuf/port_def.inc: No such file or directory
   10 | #include <google/protobuf/port_def.inc>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.

then got a whole raft of errors, too many to list.

My hope at the start of all of this was that I might be able to action my own feature request, to remove some burden from you guys, but I think I'm actually causing more pain that just waiting for the feature request and the new ubuntu deb file.

I'm happy to close this call and just wait for the fix. I think it will be easier on all concerned - especially you guys. 
Up to you if you can put up with a c++ newbie (albeit with 31 years of dev experience. I have to admit these compiling issues are exactly why I gave up trying C++ and other compiled languages years ago in favour of scripting languages.)

**Comment by albmed** [comment]
Hi @VinceLewis,

I don't understand why you did copy protobuf includes in `/usr/local/include`. It's not in the windows wiki and it's worth nothing.  

Let me explain: you are building a windows executable from a linux system. So, anything you have in your linux system is useless, because libraries, includes, executables will create objects and binaries for Linux. That's why you need MXE environment. It has all tools necessary to build windows executables in a Linux system and those are the one you must use. Neither `make`, nor `qmake` or `protoc` or libraries from your Linux system are useful. All commands must be the ones provided by MXE. 
**So, read carefully the wiki to build binaries for Windows system.** 

To sum up: 

- Install MXE and build all libraries necessary within MXE (qt5, gcc, boost, protobuf, etc.) 
- Modify `PATH` environment variable for current session. So, it can access to all tools from MXE. 
- run qmake (from MXE) 
- run make (from MXE) 

If you have problems with probuf files, you can run `protoc </path/to/pokerth>/pokerth.proto --cpp_out=</path/to/pokerth>/src/third_party/protobuf `. Of course, run `protoc` from MXE. 

I hope this clarifies. If not, don't hesitate to ask, but first read wiki ;) 

PS: About scripting languages, there are several opinions. Personally, I hate them (well, maybe except bash), but my college @q4z1 loves them :) 

Regards.

**Comment by VinceLewis** [comment]
Hi,
I have read the wiki.
I created a small script to do the windows build, this is it. Note I did
change the path includes to match what's actually on my system:

export POKERTH_DIR=/home/vince/Documents/pokerth-1.1.2-rc
export PATH=/opt/mxe/usr/x86_64-pc-linux-gnu/bin:/opt/mxe/usr/bin:$PATH
export TGT='i686-w64-mingw32.static'
rm -rf $POKERTH_DIR.$TGT.build
mkdir $POKERTH_DIR.$TGT.build
cd $POKERTH_DIR.$TGT.build
/opt/mxe/usr/bin/$TGT-qmake-qt4 $POKERTH_DIR/pokerth.pro
DEFINES+="BOOST_THREAD_USE_LIB" CONFIG+="client"
make release


Output from this build script attached. I ran it under sudo.

NB I used to love compiling RPG on an IBM AS/400 (now called an Iseries I
think). I'd love compiling again if I ever get one to work :-)
Vince

**Comment by albmed** [comment]
Hi @VinceLewis, 

These commands looks ok to me. Do you still have errors?  

Since there are a lot of messages in this issue and I'm not sure which one is the current problem, can you tell us again where are you stuck? 
If you still have problems with protobuf files, try this before running `qmake` and `make` (see screenshot attached): 

![image](https://user-images.githubusercontent.com/20071620/87407548-9a395f80-c5c2-11ea-9969-6ed2aaccdfb3.png)

1. remove files from `src/third_party/protobuf`
2. modify PATH to include MXE system 
3. run `protoc` (from MXE) 

Once created `*.pb.*` files everything should work. 
Just one more question. Where did you get the source? Did you switch to stable branch? I've only asked because, if you did, you should never see warnings such as 

> Please use 'syntax = "proto2";' or 'syntax = "proto3";' to specify a syntax version. 

Although it's just a warning, it was corrected some time ago in stable branch: https://github.com/pokerth/pokerth/blob/stable/pokerth.proto#L1 

Regards.

**Comment by q4z1** [comment]
Outdated.
